File: app/src/main/java/com/example/waveplayer/media_controller/MediaPlayerWUri.java
package com.example.waveplayer.media_controller;

import android.content.Context;
import android.media.MediaPlayer;
import android.os.Build;
import android.util.Log;

import com.example.waveplayer.random_playlist.AudioUri;

import static android.media.MediaPlayer.SEEK_CLOSEST;

public class MediaPlayerWUri {

    private static final String TAG = "MediaPlayerWURI";

    public static final Object lock = new Object();

    private MediaPlayer mediaPlayer;

    final Long id;

    private volatile boolean isPrepared = true;

    public boolean isPrepared() {
        synchronized (lock) {
            return isPrepared;
        }
    }

    private volatile boolean shouldPlay = false;

    private MediaPlayer.OnErrorListener onErrorListener;

    MediaPlayerWUri(final Context context, MediaPlayer mediaPlayer, final Long songID) {
        this.mediaPlayer = mediaPlayer;
        this.id = songID;
        mediaPlayer.setOnPreparedListener(null);
        mediaPlayer.setOnErrorListener(null);
        MOnPreparedListener mOnPreparedListener = new MOnPreparedListener(this);
        mediaPlayer.setOnPreparedListener(mOnPreparedListener);
        onErrorListener = (mediaPlayer1, i, i1) -> {
            synchronized (lock) {
                MediaController mediaController = MediaController.getInstance(context);
                mediaController.releaseMediaPlayers();
                mediaController.addToQueue(songID);
                if(!mediaController.isSongInProgress()) {
                    mediaController.playNext(context);
                }
                return false;
            }
        };
        mediaPlayer.setOnErrorListener(onErrorListener);
    }

    public void shouldPlay(boolean shouldPlay) {
        synchronized (lock) {
            if (shouldPlay && isPrepared) {
                mediaPlayer.start();
            } else {
                this.shouldPlay = shouldPlay;
            }
        }
    }

    public void release() {
        synchronized (lock) {
            setOnCompletionListener(null);
            isPrepared = false;
            shouldPlay = false;
            mediaPlayer.reset();
            mediaPlayer.release();
        }
    }

    public void prepareAsync() {
        Log.v(TAG, "Waiting for prepareAsync");
        synchronized (lock) {
            isPrepared = false;
            mediaPlayer.prepareAsync();
        }
    }

    public int getCurrentPosition() {
        synchronized (lock) {
            if (isPrepared) {
                return mediaPlayer.getCurrentPosition();
            } else {
                return -1;
            }
        }
    }

    public boolean isPlaying() {
        synchronized (lock) {
            if (isPrepared) {
                return mediaPlayer.isPlaying();
            } else {
                return false;
            }
        }
    }

    public void stop(final Context context, AudioUri audioUri) {
        Log.v(TAG, "MediaPlayer stopped");
        synchronized (lock) {
            isPrepared = false;
            shouldPlay = false;
            mediaPlayer.stop();
            resetIfMKV(audioUri, context);
        }
    }

    public void pause() {
        synchronized (lock) {
            if (isPrepared) {
                mediaPlayer.pause();
            }
            shouldPlay = false;
        }
    }

    public void seekTo(final Context context, AudioUri audioUri, int millis) {
        synchronized (lock) {
            if (isPrepared) {
                if (resetIfMKV(audioUri, context)) {
                    this.shouldPlay = true;
                    return;
                }
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                mediaPlayer.seekTo(millis, SEEK_CLOSEST);
            } else {
                mediaPlayer.seekTo(millis);
            }
        }
    }

    public void setOnCompletionListener(MediaPlayer.OnCompletionListener onCompletionListener) {
        mediaPlayer.setOnCompletionListener(onCompletionListener);
    }

    public boolean resetIfMKV(AudioUri audioUri, Context context) {
        String[] s = audioUri.displayName.split("\\.");
        if (s.length > 0) {
            if (s[s.length - 1].toLowerCase().equals("mkv")) {
                MediaController mediaController = MediaController.getInstance(context);
                mediaPlayer.reset();
                mediaPlayer.release();
                mediaPlayer = null;
                mediaPlayer = MediaPlayer.create(context, audioUri.getUri());
                mediaPlayer.setOnPreparedListener(null);
                mediaPlayer.setOnErrorListener(null);
                MOnPreparedListener mOnPreparedListener = new MOnPreparedListener(
                        mediaController.getMediaPlayerWUri(audioUri.id)
                );
                mediaPlayer.setOnPreparedListener(mOnPreparedListener);
                onErrorListener = (mediaPlayer1, i, i1) -> {
                    synchronized (MediaPlayerWUri.lock) {
                        mediaController.releaseMediaPlayers();
                        mediaController.addToQueue(audioUri.id);
                        if(!mediaController.isSongInProgress()) {
                            mediaController.playNext(context);
                        }
                        return false;
                    }
                };
                mediaPlayer.setOnErrorListener(onErrorListener);
                mediaPlayer.setOnCompletionListener(mediaController.onCompletionListener);
                return true;
            }
        }
        return false;
    }

    public class MOnPreparedListener implements MediaPlayer.OnPreparedListener {

        private final MediaPlayerWUri mediaPlayerWURI;

        public MOnPreparedListener(MediaPlayerWUri mediaPlayerWURI) {
            this.mediaPlayerWURI = mediaPlayerWURI;
        }

        @Override
        public void onPrepared(MediaPlayer mediaPlayer) {
            synchronized (mediaPlayerWURI) {
                Log.v(TAG, "MediaPlayer prepared");
                mediaPlayerWURI.isPrepared = true;
                if (shouldPlay) {
                    mediaPlayer.start();
                    shouldPlay = false;
                }
            }
        }
    }

}

File: app/src/main/java/com/example/waveplayer/media_controller/MediaPlayerWUri.kt
package com.example.waveplayer.media_controller

import android.content.*
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import android.media.MediaPlayer.OnPreparedListener
import android.os.Build
import com.example.waveplayer.random_playlist.AudioUri
import com.example.waveplayer.random_playlist.SongQueue
import java.util.*

class MediaPlayerWUri constructor(
        context: Context,
        private var mediaPlayer: MediaPlayer,
        val id: Long
        ) {

    @Volatile
    private var isPrepared: Boolean = false
    fun isPrepared(): Boolean {
        synchronized(lock) { return isPrepared }
    }

    @Volatile
    private var shouldPlay: Boolean = false
    private var onErrorListener: MediaPlayer.OnErrorListener?

    fun shouldPlay(shouldPlay: Boolean) {
        synchronized(lock) {
            if (shouldPlay && isPrepared) {
                mediaPlayer.start()
            } else {
                this.shouldPlay = shouldPlay
            }
        }
    }

    fun release() {
        synchronized(lock) {
            setOnCompletionListener(null)
            isPrepared = false
            shouldPlay = false
            mediaPlayer.reset()
            mediaPlayer.release()
        }
    }

    fun prepareAsync() {
        synchronized(lock) {
            isPrepared = false
            mediaPlayer.prepareAsync()
        }
    }

    fun getCurrentPosition(): Int {
        synchronized(lock) {
            return if (isPrepared) {
                mediaPlayer.currentPosition
            } else {
                -1
            }
        }
    }

    fun isPlaying(): Boolean {
        synchronized(lock) {
            return if (isPrepared) {
                mediaPlayer.isPlaying
            } else {
                false
            }
        }
    }

    fun stop(context: Context, audioUri: AudioUri) {
        synchronized(lock) {
            isPrepared = false
            shouldPlay = false
            mediaPlayer.stop()
            resetIfMKV(audioUri, context)
        }
    }

    fun pause() {
        synchronized(lock) {
            if (isPrepared) {
                mediaPlayer.pause()
            }
            shouldPlay = false
        }
    }

    fun seekTo(context: Context, audioUri: AudioUri, millis: Int) {
        synchronized(lock) {
            if (isPrepared) {
                if (resetIfMKV(audioUri, context)) {
                    shouldPlay = true
                    return
                }
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                mediaPlayer.seekTo(millis.toLong(), MediaPlayer.SEEK_CLOSEST)
            } else {
                mediaPlayer.seekTo(millis)
            }
        }
    }

    fun setOnCompletionListener(onCompletionListener: OnCompletionListener?) {
        mediaPlayer.setOnCompletionListener(onCompletionListener)
    }

    fun resetIfMKV(audioUri: AudioUri, context: Context): Boolean {
        val s: Array<String> = audioUri.displayName.split("\\.").toTypedArray()
        if (s.isNotEmpty()) {
            if ((s[s.size - 1].toLowerCase(Locale.ROOT) == "mkv")) {
                val mediaController: MediaController = MediaController.getInstance(context)
                val queue = SongQueue.getInstance()
                mediaPlayer.reset()
                mediaPlayer.release()
                mediaPlayer = MediaPlayer.create(context, audioUri.getUri())
                mediaPlayer.setOnPreparedListener(null)
                mediaPlayer.setOnErrorListener(null)
                mediaPlayer.setOnPreparedListener(mOnPreparedListener)
                onErrorListener = MediaPlayer.OnErrorListener { _: MediaPlayer, _: Int, _: Int ->
                    synchronized(lock) {
                        mediaController.releaseMediaPlayers()
                        queue.addToQueue(audioUri.id)
                        if (!mediaController.isSongInProgress()) {
                            mediaController.playNext()
                        }
                        return@OnErrorListener false
                    }
                }
                mediaPlayer.setOnErrorListener(onErrorListener)
                mediaPlayer.setOnCompletionListener(mediaController.onCompletionListener)
                return true
            }
        }
        return false
    }

    private val mOnPreparedListener = OnPreparedListener { mediaPlayer ->
        synchronized(this) {
            isPrepared = true
            if (shouldPlay) {
                mediaPlayer.start()
                shouldPlay = false
            }
        }
    }

    companion object {
        val lock: Any = Any()
    }

    init {
        mediaPlayer.setOnPreparedListener(null)
        mediaPlayer.setOnErrorListener(null)
        mediaPlayer.setOnPreparedListener(mOnPreparedListener)
        onErrorListener = MediaPlayer.OnErrorListener { _: MediaPlayer?, _: Int, _: Int ->
            synchronized(lock) {
                val mediaController: MediaController = MediaController.getInstance(context)
                val queue = SongQueue.getInstance()
                mediaController.releaseMediaPlayers()
                queue.addToQueue(id)
                if (!mediaController.isSongInProgress()) {
                    mediaController.playNext()
                }
                return@OnErrorListener false
            }
        }
        mediaPlayer.setOnErrorListener(onErrorListener)
    }
}